### ISSUE TITLE
No way to tell which is the boot disk is

### ISSUE THREAD

**Issue by david-crespo** [oxidecomputer/omicron#1417]
In the console we currently rely on the order of disks to mark out the boot disk for an instance. The boot disk is supposed to be the first disk (or at least I thought it was — will try to find relevant discussion). However, the disks come back in alphabetical order, which makes our UI wrong.

![2022-07-13-disk-order-bug](https://user-images.githubusercontent.com/3612203/178801749-7700c4da-4886-4941-94bd-2b7d65cef252.gif)

We are explicitly sorting instance disks by name.

https://github.com/oxidecomputer/omicron/blob/9ec389fda4a31fb8c27734c6fd6cce5845a6822d/nexus/src/db/datastore.rs#L1689

I didn't call this issue "the order is wrong" because ultimately that's not a very good API for this and I would rather have a more reliable property on the response. Maybe this should be stored as part of the disk state: in addition to the ID of the instance it's a attached to, a boolean indicating whether it is the boot disk.

I don't have a strong opinion on how important it is to say which disk is the boot disk — if it's not that important, it would be a lot easier not to bother making the distinction in the UI than it would be to make the necessary API changes to reliably make said distinction.

**Comment by davepacheco** [comment]
Even if we add a property like that, I think there _is_ an actual ordering to the disks (reflecting the order presented to the OS) and we may want the API to list them in that order (and maybe let you change it?)

**Comment by pfmooney** [comment]
Yeah. Given how OSes can be grumpy about disks moving around, I think it would need to be sparse.

**Comment by jmpesp** [comment]
@david-crespo what's the `InstanceCreate` body for the animated gif repro?

**Comment by jmpesp** [comment]
> I assume being first here is what makes boot-disk the boot disk?

Correct, yeah.

**Comment by askfongjojo** [comment]
This will be confusing for console users. Perhaps change the UI to show just a list of disks attached rather than showing something potentially incorrect? (I thought instance create did something wrong when I saw the data volume got listed as the boot disk of a 2-disk test VM I made.)

**Comment by david-crespo** [comment]
I think we can get the API to tell us which is the boot disk, whether through a property (maybe even on the instance response — `boot_disk_id`) or through the order of the list.

**Comment by leftwo** [comment]
Knowing the actual boot disk, vs. the first disk presented to the OS may not be something we can determine.
If the first disk we present to the OS is not bootable, but the 2nd disk is, I believe (and ignore me here if this is wrong)
the OS will boot from the 2nd disk.  This could result in confusion on the users side as we told them disk "A" was their
boot disk, but the OS booted from disk "B".

**Comment by david-crespo** [comment]
When that happens, can Propolis not tell Nexus which disk was the one that booted, like as part of the instance state?

**Comment by jclulow** [comment]
It's complicated.  If there is a possibility that the firmware will boot a disk other than the one in the first slot, the EFI firmware in the guest would need to communicate that out to the hypervisor somehow.  I don't believe it does anything like that today.

**Comment by david-crespo** [comment]
In that case then yeah, Angela is clearly right that being wrong is worse than saying nothing. Users can use names to differentiate for themselves, and later tags.

**Comment by askfongjojo** [comment]
We can probably assume users will name their disks with something meaningful (instead of disk1, disk2, disk3). E.g. 
<img width="1194" alt="image" src="https://user-images.githubusercontent.com/10874909/227847492-59f90e2e-709c-42b1-bd27-9d3835e590e4.png">
It'll be fine to just list all the disks attached and user will likely know which one is which. (the image also explains why I was rather surprised!)

**Comment by askfongjojo** [comment]
Taking this off the MVP board (it was tagged for MVP because of the potentially wrong info shown in console, not because of the need to distinguish the boot disk).

**Comment by david-crespo** [comment]
Closed by https://github.com/oxidecomputer/omicron/pull/6585
